File: ts/src/alias.ts
/* Copyright (c) 2021-2026 Richard Rodger, MIT License */


import type { Val } from './type'

import { makeNilErr } from './err'

import { cmpCodePoint } from './keyorder'
import { spreadSnapKey } from './val/MapVal'
import { ALIAS_NAME, aliasSetItems } from './aliasname'


const ALIAS_DECL_RE = new RegExp('(?:^|[\\s{[:,(])(' + ALIAS_NAME + ')[ \\t]*(?::|=(?!=))', 'g')
const NON_CODE_RE = /"(?:\\[\s\S]|[^"\\])*(?:"|$)|'(?:\\[\s\S]|[^'\\])*(?:'|$)|`(?:\\[\s\S]|[^`\\])*(?:`|$)|\/\/[^\n]*|#[^\n]*|\/\*[\s\S]*?(?:\*\/|$)/g

// Whether the head IS a set is aliasSetItems's answer, not the shape's.
const ALIAS_TAKE_RE = /^(\{[^}]*\})[ \t]*=[ \t]*@[ \t]*"([^"]*)"/


type AliasBinding = {
  name: string
  row: number
  col: number
  decl: string
  from: string    // the include a destructure took it from, '' if local
}


// THE NAMES A FILE BINDS, from its TEXT rather than its tree: an editor
// asks while the document is half-written and would not parse.
function aliasScope(src: string): AliasBinding[] {
  const out: AliasBinding[] = []
  const lines = src.split('\n')
  const code = src.replace(NON_CODE_RE, text => text.replace(/[^\n]/g, ' ')).split('\n')

  for (let li = 0; li < lines.length; li++) {
    const line = lines[li]
    const lead = line.length - line.replace(/^[ \t]+/, '').length
    const rest = line.substring(lead)
    const decl = line.trim()

    const tm = ALIAS_TAKE_RE.exec(rest)
    const binds = null == tm ? undefined : aliasSetItems(tm[1])
    if (null == tm || undefined === binds) {
      for (const dm of code[li].matchAll(ALIAS_DECL_RE)) {
        const col = dm.index + dm[0].indexOf(dm[1]) + 1
        out.push({ name: dm[1], row: li + 1, col, decl, from: '' })
      }
      continue
    }

    // The column is each item's own, so a set jumps to the name asked
    // for rather than to the pattern.
    let at = lead
    for (const b of binds) {
      at = line.indexOf(b.local, at)
      out.push({ name: b.local, row: li + 1, col: at + 1, decl, from: tm[2] })
      at += b.local.length
    }
  }

  return out
}


// EVERY ALIAS REFERENCE NAMES A DECLARED NAME, whether or not anything
// reaches it. Resolution is lazy, so a reference inside a template that
// nothing instantiates is never tried and a misspelling compiles clean.
// Whether a NAME is declared does not depend on what the tree holds, so
// it is answered here instead. See docs/design/ALIASES.0.md
function aliasErrors(ctx: any, root: Val): void {
  if (true !== (root as any).isMap) {
    return
  }
  const declared = new Set<string>((root as any).aliasKeys)
  const seen = new Set<Val>()

  const visit = (v: any): void => {
    if (null == v || true !== v.isVal || seen.has(v)) {
      return
    }
    seen.add(v)

    if (true === v.isRef) {
      const key: string | undefined = v.aliasKey
      if (undefined !== key && !declared.has(key)) {
        ctx.adderr(makeNilErr(ctx, 'no_path', v, undefined, 'resolve'))
      }
      return
    }

    if (true === v.isMap) {
      for (const k of Object.keys(v.peg)) {
        visit(v.peg[k])
      }
    }
    else if (Array.isArray(v.peg)) {
      for (const e of v.peg) {
        visit(e)
      }
    }
    else if (null != v.peg && true === v.peg.isVal) {
      visit(v.peg)
    }

    if ((true === v.isMap || true === v.isList) && null != v.spread.cj) {
      visit(v.spread.cj)
    }
  }

  visit(root)
} /* node:coverage ignore next 3 */


// T-1 (ALIASES.0.md sections 7 and 9). Expansion TERMINATES -- no
// parameters, no recursion, a finite name set -- but a name that names
// names expands to the product of what they hold. The budget is on
// EXPANDED SIZE and charged here, ahead of evaluation.
function aliasBudget(ctx: any, root: Val): Val | undefined {
  // A DOCUMENT THAT INCLUDES parses to a conjunct, not a map: the
  // deferred terms are where an included file's names arrive, so the
  // budget must see all of them, not only the first.
  const maps: any[] = []
  const gather = (v: any): void => {
    if (true === v?.isMap) {
      maps.push(v)
    }
    else if (true === v?.isConjunct && Array.isArray(v.peg)) {
      for (const t of v.peg) {
        gather(t)
      }
    }
  }
  gather(root)
  if (0 === maps.length) {
    return undefined
  }
  const decl: Record<string, Val> = {}
  for (const m of maps) {
    for (const k of m.aliasKeys) {
      decl[k] = m.peg[k]
    }
  }
  const limit: number = ctx.budget.alias
  const size = new Map<string, number>()
  const open = new Set<string>()
  let over: Val | undefined = undefined

  // A cycle is refused at resolution, which has not run yet, so a name
  // already open costs nothing here rather than looping.
  const nameSize = (key: string): number => {
    if (size.has(key)) {
      return size.get(key) as number
    }
    if (open.has(key) || !(key in decl)) {
      return 0
    }
    open.add(key)
    const n = valSize(decl[key])
    open.delete(key)
    size.set(key, n)
    return n
  }

  const valSize = (v: any): number => {
    if (true === v.isRef) {
      const key: string | undefined = v.aliasKey
      return undefined === key ? 1 : 1 + nameSize(key)
    }
    let n = 1
    if (true === v.isMap) {
      for (const k of Object.keys(v.peg)) {
        n += valSize(v.peg[k])
      }
    }
    else if (Array.isArray(v.peg)) {
      for (const e of v.peg) {
        n += valSize(e)
      }
    }
    else if (null != v.peg && true === v.peg.isVal) {
      n += valSize(v.peg)
    }
    if ((true === v.isMap || true === v.isList) && null != v.spread?.cj) {
      n += valSize(v.spread.cj)
    }
    return limit < n ? limit + 1 : n
  }

  let total = 0
  for (const m of maps) {
    for (const k of Object.keys(m.peg)) {
      if (!(k in decl)) {
        total += valSize(m.peg[k])
        if (limit < total) {
          break
        }
      }
    }
  }

  if (limit < total) {
    over = makeNilErr(ctx, 'alias_budget', root, undefined, 'resolve')
    ; (over as any).details = { budget: '' + limit }
  }
  return over
}


function expandAliases(root: Val, snapmap: Map<string, Val>): void {
  if (true !== (root as any).isMap) {
    return
  }

  const seen = new Set<Val>()

  const visit = (v: any, stack: string[]): void => {
    if (null == v || true !== v.isVal || seen.has(v)) {
      return
    }
    seen.add(v)

    if (true === v.isRef) {
      const key: string | undefined = v.aliasKey
      if (undefined === key) {
        return
      }
      v.expansion = undefined
      if (stack.includes(key)) {
        return
      }
      const target: Val | undefined =
        snapmap.get(spreadSnapKey(v)) ?? (root as any).peg[key]
      if (null == target) {
        return
      }
      v.expansion = target
      visit(target, [...stack, key])
      return
    }

    if (true === v.isMap) {
      // A declaration is reached through its references, each under
      // its own name, never as a child: a self-reference inside it
      // is a knot only from inside.
      const keys = Object.keys(v.peg)
        .filter((k: string) => !v.aliasKeys.includes(k))
        .sort(cmpCodePoint)
      for (const k of keys) {
        visit(v.peg[k], stack)
      }
    }
    else if (Array.isArray(v.peg)) {
      for (const e of v.peg) {
        visit(e, stack)
      }
    }
    else if (null != v.peg && true === v.peg.isVal) {
      visit(v.peg, stack)
    }

    if ((true === v.isMap || true === v.isList) && null != v.spread.cj) {
      visit(v.spread.cj, stack)
    }
  }

  visit(root, [])
} /* node:coverage ignore next 13 */


export {
  aliasBudget,
  aliasErrors,
  aliasScope,
  expandAliases,
}


export type {
  AliasBinding,
}

File: go/alias.go
/* Copyright (c) 2026 Richard Rodger, MIT License */

package aontu

import "sort"

// The default expansion budget, raised by trust.budget.alias.
const maxAliasNodes = 1000000

// T-1 (ALIASES.0.md sections 7 and 9). Expansion TERMINATES -- no
// parameters, no recursion, a finite name set -- but a name that names
// names expands to the product of what they hold. The budget is on
// EXPANDED SIZE and charged here, ahead of evaluation.
func aliasBudget(ctx *Ctx, root Val) error {
	// A DOCUMENT THAT INCLUDES parses to a conjunct, not a map: the
	// deferred terms are where an included file's names arrive.
	maps := []*MapVal{}
	var gather func(v Val)
	gather = func(v Val) {
		switch n := v.(type) {
		case *MapVal:
			maps = append(maps, n)
		case *ConjunctVal:
			for _, t := range n.peg {
				gather(t)
			}
		}
	}
	gather(root)
	if 0 == len(maps) {
		return nil
	}

	decl := map[string]Val{}
	for _, m := range maps {
		for _, k := range m.aliasKeys {
			decl[k] = m.peg[k]
		}
	}

	limit := ctx.budgetAlias
	if 0 == limit {
		limit = maxAliasNodes
	}
	size := map[string]int{}
	open := map[string]bool{}

	var valSize func(v Val) int
	// A cycle is refused at resolution, which has not run yet, so a name
	// already open costs nothing here rather than looping.
	nameSize := func(key string) int {
		if n, ok := size[key]; ok {
			return n
		}
		if open[key] {
			return 0
		}
		d, ok := decl[key]
		if !ok {
			return 0
		}
		open[key] = true
		n := valSize(d)
		delete(open, key)
		size[key] = n
		return n
	}

	valSize = func(v Val) int {
		if nil == v {
			return 0
		}
		if rv, ok := v.(*RefVal); ok {
			if key, named := rv.aliasKey(); named {
				return 1 + nameSize(key)
			}
			return 1
		}
		n := 1
		switch t := v.(type) {
		case *MapVal:
			for _, k := range t.keys {
				n += valSize(t.peg[k])
			}
			if nil != t.spread {
				n += valSize(t.spread)
			}
		case *ListVal:
			for _, e := range t.peg {
				n += valSize(e)
			}
			if nil != t.spread {
				n += valSize(t.spread)
			}
		case *ConjunctVal:
			for _, e := range t.peg {
				n += valSize(e)
			}
		case *DisjunctVal:
			for _, e := range t.peg {
				n += valSize(e)
			}
		}
		if limit < n {
			return limit + 1
		}
		return n
	}

	total := 0
	for _, m := range maps {
		for _, k := range m.keys {
			if _, isDecl := decl[k]; isDecl {
				continue
			}
			total += valSize(m.peg[k])
			if limit < total {
				break
			}
		}
	}

	if limit < total {
		en := newNil("alias_budget")
		en.details = map[string]string{"budget": itoa(limit)}
		en.sp = root.pos()
		ctx.adderr(en)
		return &AontuError{Msg: ctx.errmsg(), Code: "alias_budget",
			Details: en.details}
	}
	return nil
}

// EVERY ALIAS REFERENCE NAMES A DECLARED NAME, whether or not anything
// reaches it. Resolution is lazy, so a reference inside a template that
// nothing instantiates is never tried and a misspelling compiles clean.
// Whether a NAME is declared does not depend on what the tree holds, so
// it is answered here instead. See docs/design/ALIASES.0.md
func aliasErrors(ctx *Ctx, root Val) error {
	rm, ok := root.(*MapVal)
	if !ok {
		return nil
	}
	declared := map[string]bool{}
	for _, k := range rm.aliasKeys {
		declared[k] = true
	}

	seen := map[Val]bool{}
	bad := 0

	var visit func(v Val)
	visit = func(v Val) {
		if nil == v || seen[v] {
			return
		}
		seen[v] = true

		switch n := v.(type) {
		case *RefVal:
			key, isAlias := n.aliasKey()
			if isAlias && !declared[key] {
				makeNilErrFull(ctx, "no_path", n, nil, "resolve", nil)
				bad++
			}

		case *MapVal:
			for _, k := range n.keys {
				visit(n.peg[k])
			}
			if nil != n.spread {
				visit(n.spread)
			}

		case *ListVal:
			for _, e := range n.peg {
				visit(e)
			}
			if nil != n.spread {
				visit(n.spread)
			}

		case *ConjunctVal:
			for _, e := range n.peg {
				visit(e)
			}

		case *DisjunctVal:
			for _, e := range n.peg {
				visit(e)
			}

		case *FuncVal:
			for _, e := range n.peg {
				visit(e)
			}

		case *PlusOpVal:
			for _, e := range n.peg {
				visit(e)
			}
		}
	}

	visit(root)

	if 0 == bad {
		return nil
	}
	return &AontuError{Msg: ctx.errmsg(), Code: "no_path"}
}

func expandAliases(root Val, snapmap map[string]Val) {
	rm, ok := root.(*MapVal)
	if !ok {
		return
	}

	seen := map[Val]bool{}

	var visit func(v Val, stack []string)
	visit = func(v Val, stack []string) {
		if nil == v || seen[v] {
			return
		}
		seen[v] = true

		switch n := v.(type) {
		case *RefVal:
			key, isAlias := n.aliasKey()
			if !isAlias {
				return
			}
			n.expansion = nil
			for _, s := range stack {
				if s == key {
					return
				}
			}
			target, snapped := snapmap[refSnapKey(n)]
			if !snapped {
				target = rm.peg[key]
			}
			if nil == target {
				return
			}
			n.expansion = target
			visit(target, append(append([]string{}, stack...), key))

		case *MapVal:
			// A declaration is reached through its references, each
			// under its own name, never as a child: a self-reference
			// inside it is a knot only from inside.
			keys := make([]string, 0, len(n.keys))
			for _, k := range n.keys {
				if !n.isAliasKey(k) {
					keys = append(keys, k)
				}
			}
			sort.Strings(keys)
			for _, k := range keys {
				visit(n.peg[k], stack)
			}
			if nil != n.spread {
				visit(n.spread, stack)
			}

		case *ListVal:
			for _, e := range n.peg {
				visit(e, stack)
			}
			if nil != n.spread {
				visit(n.spread, stack)
			}

		case *ConjunctVal:
			for _, e := range n.peg {
				visit(e, stack)
			}

		case *DisjunctVal:
			for _, e := range n.peg {
				visit(e, stack)
			}

		case *FuncVal:
			for _, e := range n.peg {
				visit(e, stack)
			}

		case *PlusOpVal:
			for _, e := range n.peg {
				visit(e, stack)
			}

		case *PrefVal:
			visit(n.peg, stack)

		case *ExpectVal:
			visit(n.peg, stack)
		}
	}

	visit(root, nil)
}
